Source/WebCore/loader/ResourceMonitorChecker.cpp

```diff
@@ -0,0 +1,99 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ResourceMonitorChecker.h"
+
+#include <wtf/CryptographicallyRandomNumber.h>
+#include <wtf/StdLibExtras.h>
+
+namespace WebCore {
+
+#if ENABLE(CONTENT_EXTENSIONS)
+
+ResourceMonitorChecker& ResourceMonitorChecker::singleton()
+{
+    static NeverDestroyed<ResourceMonitorChecker> globalChecker;
+    return globalChecker;
+}
+
+ResourceMonitorChecker::ResourceMonitorChecker()
+    : m_workQueue { WorkQueue::create("ResourceMonitorChecker Work Queue"_s) }
+{
+}
+
+void ResourceMonitorChecker::checkEligibility(const ContentExtensions::ResourceLoadInfo& info, CompletionHandler<void(Eligibility)>&& completionHandler)
+{
+    ASSERT(isMainThread());
+
+    ContentExtensions::ResourceLoadInfo copy = {
+        .resourceURL = info.resourceURL.isolatedCopy(),
+        .mainDocumentURL = info.mainDocumentURL.isolatedCopy(),
+        .frameURL = info.frameURL.isolatedCopy(),
+        .type = info.type,
+        .mainFrameContext = info.mainFrameContext,
+    };
+
+    protectedWorkQueue()->dispatch([this, info = WTFMove(info), completionHandler = WTFMove(completionHandler)] mutable {
+        Eligibility eligibility = Eligibility::Unsure;
+        {
+            Locker locker { m_lock };
+            eligibility = checkEligibilityWithLock(info);
+        }
+
+        callOnMainRunLoop([eligibility, completionHandler = WTFMove(completionHandler)] mutable {
+            completionHandler(eligibility);
+        });
+    });
+}
+
+ResourceMonitor::Eligibility ResourceMonitorChecker::checkEligibilityForTesting(const ContentExtensions::ResourceLoadInfo& info)
+{
+    ASSERT(isMainThread());
+
+    Locker locker { m_lock };
+    return checkEligibilityWithLock(info);
+}
+
+ResourceMonitor::Eligibility ResourceMonitorChecker::checkEligibilityWithLock(const ContentExtensions::ResourceLoadInfo& info)
+{
+    if (!m_ruleList)
+        return Eligibility::Unsure;
```

Comment:
It seems like if you have an `Unsure` state then you will have to reschedule the eligibility check some time later after the content rule list is set.

If we are pretty sure that the content rule list will be set in a timely manner, maybe it makes more sense to get rid of the lock and the `Unsure` state? For instance, add a `Vector<Function<void()> m_pendingChecks` to the class, and append to that vector if `m_ruleList` is null, otherwise enqueue the checking function to the work queue. That way when the work queue function runs it always knows there is a valid non-null rule list.

This type of change might not make sense if you allow a transition from a non-null `m_ruleList` to a null `m_ruleList` though.

Reply:
Nice. I've never heard that convenient class. I'll use it.

Reply:
I will switch implementation in next patch where actual rule list is delivered.